iroha/src/p2p.rs

```diff
@@ -0,0 +1,259 @@
+use futures::{future::FutureExt, lock::Mutex, pin_mut, prelude::*, select};
+use iroha_derive::*;
+use iroha_network::{prelude::*, Network};
+use parity_scale_codec::{Decode, Encode};
+use std::{
+    collections::{HashMap, HashSet},
+    sync::Arc,
+    time::{Duration, SystemTime},
+};
+
+use rand::{distributions::Standard, thread_rng, Rng};
+
+fn gen_random_bytes(n_bytes: usize) -> Vec<u8> {
+    let mut rng = thread_rng();
+    rng.sample_iter(&Standard).take(n_bytes).collect()
+}
+
+fn current_time() -> Duration {
+    SystemTime::now()
+        .duration_since(SystemTime::UNIX_EPOCH)
+        .expect("Failed to get duration since UNIX_EPOCH.")
+}
+
+const PING_SIZE: usize = 32;
+
+#[derive(Io, Decode, Encode, PartialEq, Debug)]
+enum Message {
+    Ping(Ping),
+    Pong(Ping),
+    PendingTx(Tx),
+    AddPeer(Peer),
+    NewPeer(Peer),
+    RemovePeer(Peer),
+}
+
+#[derive(Encode, Decode, PartialEq, Debug, Clone)]
+struct Tx {
```

Comment:
What is the difference with `tx::Transaction`?

Reply:
Yes, It should be `tx:Transaction`. I will change it.